src/etos_api/library/docker.py

```diff
@@ -0,0 +1,175 @@
+REPO_DELIMITER = "/"
+TAG_DELIMITER = ":"
+
+
+class Docker:
+    """Docker handler for HTTP operations against docker registries.
+
+    This handler is heavily inspired by `crane digest`:
+    https://github.com/google/go-containerregistry/tree/main/cmd/crane
+    """
+
+    logger = logging.getLogger(__name__)
+    # In-memory database for stored authorization tokens.
+    # This dictionary shares memory with all instances of `Docker`, by design.
+    tokens = {}
+    lock = Lock()
+
+    def token(self, manifest_url):
+        """Get a stored token, removing it if expired.
+
+        :param manifest_url: URL the token has been stored for.
+        :return: A token or None.
+        """
+        with self.lock:
+            token = self.tokens.get(manifest_url, {})
+            if token:
+                if time.time() >= token["expire"]:
+                    self.logger.info("Registry token expired for %r", manifest_url)
+                    self.tokens.pop(manifest_url)
+                    token = {}
+            return token.get("token")
+
+    async def head(
+        self, session: aiohttp.ClientSession, url: str, token: str = None
+    ) -> aiohttp.ClientResponse:
+        """Make a HEAD request to a URL, adding token to headers if supplied.
+
+        :param session: Client HTTP session to use for HTTP request.
+        :param url: URL to make HEAD request to.
+        :param token: Optional authorization token.
+        :return: HTTP response.
+        """
+        headers = {}
+        if token is not None:
+            headers["Authorization"] = f"Bearer {token}"
+
+        async with session.head(url, headers=headers) as response:
+            return response
+
+    async def authorize(
+        self, session: aiohttp.ClientSession, response: aiohttp.ClientResponse
+    ) -> str:
+        """Get a token from an unauthorized request to image repository.
+
+        :param session: Client HTTP session to use for HTTP request.
+        :param response: HTTP response to get headers from.
+        :return: Response JSON from authorization request.
+        """
+        header = response.headers.get("www-authenticate")
+        header = header.replace("Bearer ", "")
+        parts = header.split(",")
+
+        url = None
+        query = {}
+        for part in parts:
+            key, value = part.split("=")
+            if key == "realm":
+                url = value.strip('"')
+            else:
+                query[key] = value.strip('"')
+
+        async with session.get(url, params=query) as response:
+            response.raise_for_status()
+            return await response.json()
+
+    def tag(self, base: str) -> tuple[str, str]:
+        """Figure out tag from a container image name.
+
+        :param base: Name of image.
+        :return: Base image name without tag and tag name.
+        """
+        tag = ""
+
+        parts = base.split(TAG_DELIMITER)
+        if len(parts) > 1 and REPO_DELIMITER not in parts[-1]:
```

Comment:
Add explanatory comment on code. E.g all the ins ans outs of ports and missing tags etc.

Reply:
Please check the comments added.